### PR TITLE
RavenDB-17055 : Sharded ETL

### PR DESCRIPTION
**Pull request by aviv86**
https://issues.hibernatingrhinos.com/issue/RavenDB-17055

### CODE REVIEW

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 1]
```diff
@@ -675,7 +678,7 @@ public void Run()
 
                     var state  = LastProcessState = GetProcessState(Database, Configuration.Name, Transformation.Name);
 
-                    var loadLastProcessedEtag = state.GetLastProcessedEtagForNode(_serverStore.NodeTag);
+                    var loadLastProcessedEtag = state.GetLastProcessedEtag(Database.DbBase64Id);
```
Comment (ayende) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs]: Here, shouldn't we also try the `GetLastProcessedEtagForNode` if this returns `0` ? In case we have legacy data?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs [hunk 2]
```diff
@@ -0,0 +1,549 @@
...
+using Raven.Server.Utils;
+using Raven.Server.Web.System;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Documents.ShardedHandlers
+{
+    public class ShardedOngoingTasksHandler : ShardedRequestHandler
+    {
+        [RavenShardedAction("/databases/*/admin/connection-strings", "PUT")]
```
Comment (ayende) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs:26]: A lot of this looks exactly like the non sharded version, can we delagate that by calling the `OngoingTasksHandler`, instead?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs [hunk 3]
```diff
@@ -0,0 +1,549 @@
...
+                        case OngoingTaskType.Subscription:
+                        case OngoingTaskType.PullReplicationAsSink:
+                        case OngoingTaskType.Backup:
+                            // todo
+                            break;
+                        case OngoingTaskType.PullReplicationAsHub:
+                            throw new BadRequestException("Getting task info for " + OngoingTaskType.PullReplicationAsHub + " is not supported");
+
+                        case OngoingTaskType.SqlEtl:
+/*
```
Comment (ayende) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs]: why all the commented code?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 4]
```diff
@@ -1166,24 +1165,44 @@ private void RemoveNodeFromCluster(ClusterOperationContext context, BlittableJso
...
                             {
-                                DeleteDatabaseRecord(context, index, items, lowerKey, record.DatabaseName, serverStore);
+                                DeleteDatabaseRecord(context, index, items, lowerKey, record, serverStore);
+                                continue;
+                            }
+                        }
+                        else
+                        {
+                            bool emptyShardTypology = true;
+                            foreach (var shardTopology in record.Shards)
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:1199]: I think this is problematic. 
What happens when we have:

* A,B,C
* Database with a shard on each?
* We now remove the node...
* That will delete *part* of the database, however, no?


Looks like that is something that we should guard against.

File: src/Raven.Server/Documents/ETL/EtlProcess.cs (hunk 1 above)
Comment (aviv86) [reply src/Raven.Server/Documents/ETL/EtlProcess.cs]: fixed

File: src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs (hunk 2 above)
Comment (aviv86) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs:26]: done

File: src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs (hunk 3 above)
Comment (aviv86) [reply src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs]: removed

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 4 above)
Comment (aviv86) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:1199]: yes, we will handle that in a separate ticket 
https://issues.hibernatingrhinos.com/issue/RavenDB-17161

File: src/Raven.Server/Documents/ETL/EtlLoader.cs [hunk 5]
```diff
@@ -692,12 +692,15 @@ public void HandleDatabaseRecordChange(DatabaseRecord record)
 
         public void HandleDatabaseValueChanged(DatabaseRecord record)
         {
+            var dbName = record.DatabaseName;
+            ShardHelper.TryGetShardIndexAndDatabaseName(ref dbName);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ETL/EtlLoader.cs:696]: can this return 'false' ? should we throw?

File: src/Raven.Server/Documents/ETL/EtlProcess.cs [hunk 6]
```diff
@@ -89,10 +89,13 @@ public abstract class EtlProcess : IDisposable
 
         public static EtlProcessState GetProcessState(DocumentDatabase database, string configurationName, string transformationName)
         {
+            var databaseName = database.Name;
+            ShardHelper.TryGetShardIndexAndDatabaseName(ref databaseName);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/ETL/EtlProcess.cs:93]: can this return 'false' ? should we throw?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 7]
```diff
@@ -1170,24 +1169,49 @@ private void RemoveNodeFromCluster(ClusterOperationContext context, BlittableJso
...
                             // Explicit removing of the node means that we modify the replication factor
                             record.Topology.ReplicationFactor = record.Topology.Count;
                             if (record.Topology.Count == 0)
                             {
-                                DeleteDatabaseRecord(context, index, items, lowerKey, record.DatabaseName, serverStore);
+                                DeleteDatabaseRecord(context, index, items, lowerKey, record, serverStore);
+                                continue;
+                            }
+                        }
+                        else
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:1196]: is this correct `else` ? we can go here when we have `IsSharded` but also when RelevantFor is false?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 8]
```diff
@@ -1456,24 +1480,42 @@ private static void CleanupDatabaseReplicationCertificate(ClusterOperationContex
...
+                    }
+
+                    // shard can be idle when we are deleting it
+                    serverStore?.IdleDatabases.TryRemove(shard, out _);
+                }
+            }
+            else
+            {
+                // db can be idle when we are deleting it
+                serverStore?.IdleDatabases.TryRemove(record.DatabaseName, out _);
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:1511]: why is this needed?

File: src/Raven.Server/ServerWide/Commands/ETL/UpdateEtlProcessStateCommand.cs [hunk 9]
```diff
@@ -52,7 +53,10 @@ private UpdateEtlProcessStateCommand()
 
         public override string GetItemId()
         {
-            return EtlProcessState.GenerateItemName(DatabaseName, ConfigurationName, TransformationName);
+            var databaseName = DatabaseName;
+            ShardHelper.TryGetShardIndexAndDatabaseName(ref databaseName);
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/ETL/UpdateEtlProcessStateCommand.cs:57]: can this fail?

File: src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                    if (Enum.TryParse<OngoingTaskType>(typeStr, true, out var type) == false)
+                        throw new ArgumentException($"Unknown task type: {type}", "type");
+
+                    switch (type)
+                    {
+                        case OngoingTaskType.Replication:
+                        case OngoingTaskType.Subscription:
+                        case OngoingTaskType.PullReplicationAsSink:
+                        case OngoingTaskType.Backup:
+                            // todo
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs]: Minor tip that I found very useful - in addition to adding `// todo` I usually add my login e.g. `// todo arek` this way I can very easily find all my TODOs

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 11]
```diff
@@ -1456,24 +1480,42 @@ private static void CleanupDatabaseReplicationCertificate(ClusterOperationContex
...
             {
                 items.DeleteByPrimaryKeyPrefix(loweredKey);
             }
 
-            var databaseLowered = $"{databaseName.ToLowerInvariant()}/";
+            if (record.IsSharded)
+            {
+                for (int i = 0; i < record.Shards.Length; i++)
+                {
+                    var shard = record.DatabaseName + "$" + i;
```
Comment (arekpalinski) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: I believe that's something what we'll commonly do. Maybe it would be good to create some shard helper class?

**Comment by arekpalinski** [comment]
In general. Looks good to me

File: src/Raven.Server/Documents/ETL/EtlLoader.cs (hunk 5 above)
Comment (aviv86) [reply src/Raven.Server/Documents/ETL/EtlLoader.cs:696]: no, this can return -1 if this is not a sharded name with `$` suffix, and then `dbName` is not changed 
this code is going to be called from non-shrded etl too so we shouldn't throw

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 7 above)
Comment (aviv86) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:1196]: fixed

File: src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                            throw new BadRequestException("Getting task info for " + OngoingTaskType.PullReplicationAsHub + " is not supported");
+                        case OngoingTaskType.SqlEtl:
+                        case OngoingTaskType.OlapEtl:
+                        case OngoingTaskType.RavenEtl:
+                            if (name == null)
+                                throw new ArgumentException($"ETL task {key} is sharded, you must specify a query string argument for 'name', but none was specified.");
+
+                            var index = ShardHelper.TryGetShardIndexAndDatabaseName(ref name);
+                            if (index == -1)
+                                throw new ArgumentException($"ETL task '{name}' is sharded, you must specify the shard index, for example : '{name}$0'");
```
Comment (ayende) [inline src/Raven.Server/Documents/ShardedHandlers/ShardedOngoingTasks.cs:120]: What is the meaning hear, for `name$0` ? What is the user expected to do, pass the value for the specific instance?
Also, given that we are just forwarding, do we need that much code here?

File: src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs [hunk 13]
```diff
@@ -51,5 +55,24 @@ public void AddHeaders<T>(ShardedBaseCommand<T> command, Headers header)
             if (header.HasFlag(Headers.IfNoneMatch))
                 command.Headers["If-None-Match"] = GetStringFromHeaders("If-None-Match");
         }
+
+        public override async Task WaitForIndexToBeApplied(TransactionOperationContext context, long index)
+        {
+            var dbs = ServerStore.DatabasesLandlord.TryGetOrCreateShardedResourcesStore(ShardedContext.DatabaseName).ToList();
```
Comment (ayende) [inline src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs:61]: What is the meaning here? What is this for?

Comment (ayende) [reply src/Raven.Server/Documents/Sharding/ShardedRequestHandler.cs:61]: Pretty sure that this is conceptually wrong, because you may be needing to wait on a value on a _different node_

File: src/Raven.Server/Web/RequestHandler.cs [hunk 14]
```diff
@@ -778,6 +778,12 @@ protected virtual OperationCancelToken CreateOperationToken()
         protected virtual OperationCancelToken CreateOperationToken(TimeSpan cancelAfter)
         {
             return new OperationCancelToken(cancelAfter, ServerStore.ServerShutdown, HttpContext.RequestAborted);
+        }
+
+        public virtual Task WaitForIndexToBeApplied(TransactionOperationContext context, long index)
+        {
+            return Task.CompletedTask;
```
Comment (ayende) [inline src/Raven.Server/Web/RequestHandler.cs:785]: why?
